### PR TITLE
Workers dequeuing jobs from queues using both Round-Robin and Random strategies

### PR DESCRIPTION
**Pull request by bielcardona**
I have subclassed Worker so that the subclass can dequeue jobs using both a round-robin strategy or a random strategy, using the method reorder_queues().
I have had to add a call to reorder_queues in the base class, which does nothing in this case.

### CODE REVIEW

**Comment by thesamesam** [comment]
Nice! This looks really helpful. It'd also fix #1423.

**Comment by jpsca** [comment]
@bielcardona  There is something wrong with the branch. GitHub only sees as changed `rq/worker.py` but not the new tests or documentation.

**Comment by bielcardona** [comment]
I have added doc and tests. The tests run with no error locally, but it seems that the automatic tests run on github fail (not sure if due to my changes)

**Comment by thesamesam** [comment]
Hmm. This is the failure: https://github.com/rq/rq/runs/2080885359#step:7:50. I'm not sure where "In" is coming from.

(Squashing the commits may be helpful too btw)

**Comment by jpsca** [comment]
@thesamesam `float("Inf")` is how you get an "infinite" in Python, probably someone deleted the final "f" by mistake

**Comment by bielcardona** [comment]
oops... sorry. I have corrected it.

**Comment by aparcar** [comment]
I'm a bit confused with this concept, the worker first pics a job and then modifies the queue list according to it's *dequeue style*? So then the next worker pics the job according to the previous worker *reordering* and then reoders itself again?

**Comment by bielcardona** [comment]
@aparcar 
> I'm a bit confused with this concept, the worker first pics a job and then modifies the queue list according to it's dequeue style?

Yes

> So then the next worker pics the job according to the previous worker reordering and then reoders itself again?

No. Each worker has its own queue list.

**Comment by selwin** [comment]
This is a much better attempt, but instead of having two additional worker classes, wouldn't it be better to have a  `dequeue_strategy` attribute on worker?

File: rq/worker.py [hunk 1]
```diff
@@ -1160,3 +1164,22 @@ def request_force_stop_sigrtmin(self, signum, frame):
...
         raise ShutDownImminentException('shut down imminent (signal: %s)' % signal_name(signum), info)
+
+
+class RoundRobinWorker(Worker):
+    """
+    Modified version of Worker that dequeues jobs from the queues using a round-robin strategy.
+    """
+    def reorder_queues(self, reference_queue):
+        pos = self.queues.index(reference_queue)
+        self.queues=self.queues[pos+1:]+self.queues[:pos+1]
```
Comment (selwin) [inline rq/worker.py:1175]: I don't think we should mutate `self.queues`. Can you create another property `self._ordered_queues` that we use for this purpose?

**Comment by bielcardona** [comment]
> This is a much better attempt, but instead of having two additional worker classes, wouldn't it be better to have a `dequeue_strategy` attribute on worker?

Having custom worker classes makes it straightforward to use these strategies from the command line with no change to the API. You just have to create the worker with the appropriate --worker-class parameter. I think it is a simpler way to do it.

**Comment by selwin** [comment]
Ok, I have no hard opinions on this. Can you implement the change to store the reordered queues in `_ordered_queues`?

File: rq/worker.py (hunk 1 above)
Comment (selwin) [inline rq/worker.py:1175]: Can you add the proper spacing around `=` and `+`?

File: rq/worker.py [hunk 2]
```diff
@@ -1160,3 +1164,22 @@ def request_force_stop_sigrtmin(self, signum, frame):
...
+
+
+class RandomWorker(Worker):
+    """
+    Modified version of Worker that dequeues jobs from the queues using a random strategy.
+    """
+
+    def reorder_queues(self, reference_queue):
+        from random import shuffle
```
Comment (selwin) [inline rq/worker.py:1184]: Can you move this import line to import section at the beginning of file?

File: tests/test_worker.py [hunk 3]
```diff
@@ -1362,3 +1362,56 @@ def setUp(self):
...
+        w.work(burst=True)
+        start_times = []
+        for i in range(5):
+            for j in range(3):
+                job = Job.fetch('q%d_%d' % (i,j))
+                start_times.append(('q%d_%d' % (i,j), job.started_at))
+        sorted_by_time = sorted(start_times, key=lambda tup: tup[1])
+        sorted_ids = [tup[0] for tup in sorted_by_time]
+        expected = ['q%d_%d' % (i,j) for j in range(3) for i in range(5)]
+        self.assertEqual(expected, sorted_ids)
```
Comment (selwin) [inline tests/test_worker.py:1386]: To make this clear, can you make this more explicit? Do this:
```python
self.assertEqual(
    sorted_ids,
    ['1_1', '2_1', ...]
)
```
If the list is too long, I don't mind reducing the number of loops to 3 and 3.

File: tests/test_worker.py [hunk 4]
```diff
@@ -1362,3 +1362,56 @@ def setUp(self):
...
+
+
+class TestRandomWorker(RQTestCase):
+    def test_random_worker(self):
+        qs = [Queue('q%d' % i) for i in range(5)]
+
+        for i in range(5):
+            for j in range(3):
+                qs[i].enqueue(say_pid,
+                              job_id='q%d_%d' % (i,j))
```
Comment (selwin) [inline tests/test_worker.py:1396]: Can you add space after `,`? Sorry for nitpicking, but I want the code style to be consistent.

File: rq/worker.py (hunk 1 above)
Comment (bielcardona) [reply rq/worker.py:1175]: No problem about adding a new property, but I would like to know where it should be used (apart from the call to `dequeue_any` in line 650). For instance, I see no point in using it in `validate_queues` (line 241), but I am not sure if it should be used in `queue_names` (line 247) or `queue_keys` (line 251). Same for the call to `RQScheduler` in line 553 and the definitions of `refresh`(line 729) and `clean_registries` (line 1088).

Comment (bielcardona) [reply rq/worker.py]: done

File: rq/worker.py (hunk 2 above)
Comment (bielcardona) [reply rq/worker.py]: done

File: tests/test_worker.py (hunk 3 above)
Comment (bielcardona) [reply tests/test_worker.py:1388]: done

File: tests/test_worker.py (hunk 4 above)
Comment (bielcardona) [reply tests/test_worker.py]: done

File: rq/worker.py (hunk 1 above)
Comment (selwin) [reply rq/worker.py]: `_ordered_queues` is used to shuffle queue orderings. `self.queues` shouldn't be mutated.

**Comment by bielcardona** [comment]
No news on this pull request? I think I fixed all the issues.

**Comment by selwin** [comment]
Sorry for the delay. Will review again soon.

**Comment by selwin** [comment]
Thanks!

**Pull request merged**